graphql_compiler/compiler/emit_sql.py

```diff
@@ -1017,6 +1060,11 @@ def recurse(self, vertex_field: str, depth: int) -> None:
             )
 
         edge = self._sql_schema_info.join_descriptors[self._current_classname][vertex_field]
+        if not isinstance(edge, DirectJoinDescriptor):
+            raise NotImplementedError(
+                f"Edge {vertex_field} is backed by a CompositeJoinDescriptor, "
+                "so it can't be used with @recurse."
```

Comment:
I believe this is a user-facing error message, so let's try to avoid the parts of our terminology that are not suitable for querying users. How about something like this:
```suggestion
                f"Edge {vertex_field} requires a JOIN across a composite key, which is currently "
                f"not supported for use with @recurse."
```